include/gsl/joining_thread

```diff
@@ -0,0 +1,82 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#ifndef GSL_JOINING_THREAD_H
+#define GSL_JOINING_THREAD_H
+
+#include <thread>
+#include <utility>
+
+namespace gsl
+{
+/**
+ * RAII wrapper for std::thread.
+ * It's guaranteed that joining_thread will join started thread when destroyed.
+ */
+class joining_thread
+{
+public:
+    joining_thread() noexcept {};
+
+    template <class Function, class... Args>
+    explicit joining_thread(Function&& function, Args&&... args)
+        : thread(std::forward<Function>(function), std::forward<Args>(args)...)
+    {
+    }
+
+    joining_thread(const joining_thread&) = delete;
+    joining_thread& operator=(const joining_thread&) = delete;
```

Comment:
`= default` the copy constructor and copy assignment operator would be slightly better, because it would clearly indicate that this class has the same behavior as `std::thread`. These members would still end up being deleted because corresponding members of `std::thread` are deleted.